src/WebRTC/SessionDescriptionHandler.js

```diff
@@ -0,0 +1,487 @@
+   */
+  setDescription: {writable:true, value: function setDescription (sessionDescription, options, modifiers) {
+    var self = this;
+
+    // https://stackoverflow.com/questions/9847580/how-to-detect-safari-chrome-ie-firefox-and-opera-browser
+    var isFirefox = typeof InstallTrigger !== 'undefined';
+    if (!this.session.disableRenegotiation && isFirefox && this.peerConnection && this.isVideoHold(sessionDescription)) {
+      this.session.disableRenegotiation = true;
+    }
+
+    if (this.session.disableRenegotiation) {
+      this.logger.warn("The flag \"disableRenegotiation\" is set to true for this session description handler. We will not try to renegotiate.");
+      throw new SIP.Exceptions.RenegotiationError("disableRenegotiation flag set to true for this session description handler");
+    }
+
+    options = options || {};
+    if (options.peerConnectionOptions) {
+      this.initPeerConnection(options.peerConnectionOptions);
+    }
+
+    // Merge passed constraints with saved constraints and save
+    this.constraints = Object.assign(this.constraints, options.constraints);
+    this.constraints = this.checkAndDefaultConstraints(this.constraints);
+
+    modifiers = modifiers || [];
+    if (!Array.isArray(modifiers)) {
+      modifiers = [modifiers];
+    }
+    modifiers = modifiers.concat(this.modifiers);
+
+    return SIP.Utils.reducePromises(modifiers, sessionDescription)
+      .catch(function modifierError(e) {
+        self.logger.error("The modifiers did not resolve successfully");
+        self.logger.error(e);
+        throw e;
+      })
+      .then(function(sdp) {
+        var rawDescription = {
+          type: self.hasOffer('local') ? 'answer' : 'offer',
+          sdp: sdp
+        };
+        self.emit('setDescription', rawDescription);
+
+        return self.peerConnection.setRemoteDescription(new self.WebRTC.RTCSessionDescription(rawDescription));
+      })
+      .catch(function setRemoteDescriptionError(e) {
+        self.session.disableRenegotiation = true;
+        self.logger.error(e);
+        self.emit('peerConnection-setRemoteDescriptionFailed', e);
+        throw e;
+      })
+      .then(function setRemoteDescriptionSuccess() {
+        if (self.peerConnection.getRemoteStreams) {
+          self.emit('setRemoteDescription', self.peerConnection.getRemoteStreams());
+        } else {
+          // TODO: Shim this correctly for Safari
+          self.emit('setRemoteDescription', self.peerConnection.getSenders());
+        }
+        self.emit('confirmed', self);
+      });
+  }},
+
+  // Internal functions
+  createOfferOrAnswer: {writable: true, value: function createOfferOrAnswer (RTCOfferOptions, modifiers) {
+    var self = this;
+    var methodName;
+    var pc = this.peerConnection;
+
+    RTCOfferOptions = RTCOfferOptions || {};
+
+    methodName = self.hasOffer('remote') ? 'createAnswer' : 'createOffer';
+
+    return pc[methodName](RTCOfferOptions)
```

Comment:
We should check the state of the peer connection signaling state to make sure that the peer connection is in the correct state before doing createOffer/createAnswer. Firefox will throw an error if you are not in the correct state, chrome allows the call to still succeed.